track number of cpp->python exceptions thrown in torch.compile benchmark suite

bdhirsh opened the pull request:
cpp->python exceptions can be slow - **especially* if you are running pytorch with infra that tries to symbolize C++ stacktraces.

This PR attempts to:
(1) add a counter every time pybind catches a C++ exception that enters python
(2) track the value of this counter in our benchmark suite, so we can hopefully drive it to zero.

It might be the case that we have very few (or even zero) of these cases showing up in torchbench. Emprically from internal, it also appears that this can show up when torch.compiling custom ops. So if few-or-zero exceptions are encountered in torchbench, we should also add some tests for avoiding C++ exceptions in the common path of torch.compile usage with custom ops.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #131481
* #133044
* #137921



cc @albanD @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

bdhirsh committed:
track number of cpp->python exceptions thrown in torch.compile benchmark suite

[ghstack-poisoned]

bdhirsh committed:
Update on "track number of cpp->python exceptions thrown in torch.compile benchmark suite"

cpp->python exceptions can be slow - **especially* if you are running pytorch with infra that tries to symbolize C++ stacktraces.

This PR attempts to:
(1) add a counter every time pybind catches a C++ exception that enters python
(2) track the value of this counter in our benchmark suite, so we can hopefully drive it to zero.

It might be the case that we have very few (or even zero) of these cases showing up in torchbench. Emprically from internal, it also appears that this can show up when torch.compiling custom ops. So if few-or-zero exceptions are encountered in torchbench, we should also add some tests for avoiding C++ exceptions in the common path of torch.compile usage with custom ops.




[ghstack-poisoned]

bdhirsh commented on torch/csrc/jit/python/init.cpp:
```diff
@@ -166,6 +166,8 @@ std::optional<IValue> toTypeInferredIValueOptional(py::handle input) {
 TORCH_API void runJITCPPTests();
 #endif
 
+static thread_local uint64_t cpp_to_python_translated_exception_count{0};
```
My thought process here was:

(1) I figured this can't be a plain `uint64_t`, since multiple threads may be raising exceptions and can increment the counter.

(2) I originally made it a `std::atomic<uint64_t>`, but then I didn't want to risk the increment code being very slow, and exacerbating slowness problems in cases where there are many cpp-to-python exceptions. Alternatively, we could make it slow but gate it behind a config / env var, but this seems like a nice metric to be able to track in an always-on way.

(3) thread_local will under-count if multiple threads are all raising exceptions. My hope is that we are very unlikely to hit a situation where the main thread raises no cpp-to-python exception, but a side thread raises a large number (although if this happens, then... we will not be able to track it in metrics).

ezyang replied:
Don't worry about (2), you're already in slow pain with exceptions, the atomic is a rounding error.

bdhirsh committed:
Update on "track number of cpp->python exceptions thrown in torch.compile benchmark suite"

cpp->python exceptions can be slow - **especially* if you are running pytorch with infra that tries to symbolize C++ stacktraces.

This PR attempts to:
(1) add a counter every time pybind catches a C++ exception that enters python
(2) track the value of this counter in our benchmark suite, so we can hopefully drive it to zero.

It might be the case that we have very few (or even zero) of these cases showing up in torchbench. Emprically from internal, it also appears that this can show up when torch.compiling custom ops. So if few-or-zero exceptions are encountered in torchbench, we should also add some tests for avoiding C++ exceptions in the common path of torch.compile usage with custom ops.




[ghstack-poisoned]

bdhirsh commented:
I gave up looking for a way to grab the output from CI, so I just ran a script that adds the extra column to every existing csv we check in (set to zero - if there are any exceptions, I'll expect CI to fail)

bdhirsh committed:
Update on "track number of cpp->python exceptions thrown in torch.compile benchmark suite"

cpp->python exceptions can be slow - **especially* if you are running pytorch with infra that tries to symbolize C++ stacktraces.

This PR attempts to:
(1) add a counter every time pybind catches a C++ exception that enters python
(2) track the value of this counter in our benchmark suite, so we can hopefully drive it to zero.

It might be the case that we have very few (or even zero) of these cases showing up in torchbench. Emprically from internal, it also appears that this can show up when torch.compiling custom ops. So if few-or-zero exceptions are encountered in torchbench, we should also add some tests for avoiding C++ exceptions in the common path of torch.compile usage with custom ops.




[ghstack-poisoned]

bdhirsh committed:
Update on "track number of cpp->python exceptions thrown in torch.compile benchmark suite"

cpp->python exceptions can be slow - **especially* if you are running pytorch with infra that tries to symbolize C++ stacktraces.

This PR attempts to:
(1) add a counter every time pybind catches a C++ exception that enters python
(2) track the value of this counter in our benchmark suite, so we can hopefully drive it to zero.

It might be the case that we have very few (or even zero) of these cases showing up in torchbench. Emprically from internal, it also appears that this can show up when torch.compiling custom ops. So if few-or-zero exceptions are encountered in torchbench, we should also add some tests for avoiding C++ exceptions in the common path of torch.compile usage with custom ops.




[ghstack-poisoned]

bdhirsh committed:
Update on "track number of cpp->python exceptions thrown in torch.compile benchmark suite"

cpp->python exceptions can be slow - **especially* if you are running pytorch with infra that tries to symbolize C++ stacktraces.

This PR attempts to:
(1) add a counter every time pybind catches a C++ exception that enters python
(2) track the value of this counter in our benchmark suite, so we can hopefully drive it to zero.

It might be the case that we have very few (or even zero) of these cases showing up in torchbench. Emprically from internal, it also appears that this can show up when torch.compiling custom ops. So if few-or-zero exceptions are encountered in torchbench, we should also add some tests for avoiding C++ exceptions in the common path of torch.compile usage with custom ops.




[ghstack-poisoned]

bdhirsh committed:
Update on "track number of cpp->python exceptions thrown in torch.compile benchmark suite"

cpp->python exceptions can be slow - **especially* if you are running pytorch with infra that tries to symbolize C++ stacktraces.

This PR attempts to:
(1) add a counter every time pybind catches a C++ exception that enters python
(2) track the value of this counter in our benchmark suite, so we can hopefully drive it to zero.

It might be the case that we have very few (or even zero) of these cases showing up in torchbench. Emprically from internal, it also appears that this can show up when torch.compiling custom ops. So if few-or-zero exceptions are encountered in torchbench, we should also add some tests for avoiding C++ exceptions in the common path of torch.compile usage with custom ops.




[ghstack-poisoned]

bdhirsh committed:
Update on "track number of cpp->python exceptions thrown in torch.compile benchmark suite"

cpp->python exceptions can be slow - **especially* if you are running pytorch with infra that tries to symbolize C++ stacktraces.

This PR attempts to:
(1) add a counter every time pybind catches a C++ exception that enters python
(2) track the value of this counter in our benchmark suite, so we can hopefully drive it to zero.

It might be the case that we have very few (or even zero) of these cases showing up in torchbench. Emprically from internal, it also appears that this can show up when torch.compiling custom ops. So if few-or-zero exceptions are encountered in torchbench, we should also add some tests for avoiding C++ exceptions in the common path of torch.compile usage with custom ops.




[ghstack-poisoned]

bdhirsh committed:
Update on "track number of cpp->python exceptions thrown in torch.compile benchmark suite"

cpp->python exceptions can be slow - **especially* if you are running pytorch with infra that tries to symbolize C++ stacktraces.

This PR attempts to:
(1) add a counter every time pybind catches a C++ exception that enters python
(2) track the value of this counter in our benchmark suite, so we can hopefully drive it to zero.

It might be the case that we have very few (or even zero) of these cases showing up in torchbench. Emprically from internal, it also appears that this can show up when torch.compiling custom ops. So if few-or-zero exceptions are encountered in torchbench, we should also add some tests for avoiding C++ exceptions in the common path of torch.compile usage with custom ops.




[ghstack-poisoned]

bdhirsh commented:
Interesting, there are now zero exceptions in the test suite except for `dlrm`, which has 1

bdhirsh committed:
Update on "track number of cpp->python exceptions thrown in torch.compile benchmark suite"

cpp->python exceptions can be slow - **especially* if you are running pytorch with infra that tries to symbolize C++ stacktraces.

This PR attempts to:
(1) add a counter every time pybind catches a C++ exception that enters python
(2) track the value of this counter in our benchmark suite, so we can hopefully drive it to zero.

It might be the case that we have very few (or even zero) of these cases showing up in torchbench. Emprically from internal, it also appears that this can show up when torch.compiling custom ops. So if few-or-zero exceptions are encountered in torchbench, we should also add some tests for avoiding C++ exceptions in the common path of torch.compile usage with custom ops.




[ghstack-poisoned]

bdhirsh commented:
I ran the command to pull results from CI and overwrite them locally and re-pushed. interestingly:

(1) there are no exceptions (including in dlrm) according to the results

(2) CI still fails though: the overwrite removed the new column from a few of the expected csv files, and CI now complains when trying to compare them

ezyang reviewed (approved):
Don't stress too much about the csv bits, if you need to do something more manual for a handful of models do that

albanD reviewed (commented):
I'm a bit confused, you never actually increment this counter so its forever 0?
The next PR doesn't seem to add increments either?

EDIT: From offline discussion, the goal is NOT to catch all cpp->python exceptions (which would require hooking in Exceptions.h where we do most of these) but catch when pybind exception handler is triggered. Which this PR does!

bdhirsh committed:
Update on "track number of cpp->python exceptions thrown in torch.compile benchmark suite"

cpp->python exceptions can be slow - **especially* if you are running pytorch with infra that tries to symbolize C++ stacktraces.

This PR attempts to:
(1) add a counter every time pybind catches a C++ exception that enters python
(2) track the value of this counter in our benchmark suite, so we can hopefully drive it to zero.

It might be the case that we have very few (or even zero) of these cases showing up in torchbench. Emprically from internal, it also appears that this can show up when torch.compiling custom ops. So if few-or-zero exceptions are encountered in torchbench, we should also add some tests for avoiding C++ exceptions in the common path of torch.compile usage with custom ops.




[ghstack-poisoned]

bdhirsh committed:
Update on "track number of cpp->python exceptions thrown in torch.compile benchmark suite"

cpp->python exceptions can be slow - **especially* if you are running pytorch with infra that tries to symbolize C++ stacktraces.

This PR attempts to:
(1) add a counter every time pybind catches a C++ exception that enters python
(2) track the value of this counter in our benchmark suite, so we can hopefully drive it to zero.

It might be the case that we have very few (or even zero) of these cases showing up in torchbench. Emprically from internal, it also appears that this can show up when torch.compiling custom ops. So if few-or-zero exceptions are encountered in torchbench, we should also add some tests for avoiding C++ exceptions in the common path of torch.compile usage with custom ops.




[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge

bdhirsh committed:
Update on "track number of cpp->python exceptions thrown in torch.compile benchmark suite"

cpp->python exceptions can be slow - **especially* if you are running pytorch with infra that tries to symbolize C++ stacktraces.

This PR attempts to:
(1) add a counter every time pybind catches a C++ exception that enters python
(2) track the value of this counter in our benchmark suite, so we can hopefully drive it to zero.

It might be the case that we have very few (or even zero) of these cases showing up in torchbench. Emprically from internal, it also appears that this can show up when torch.compiling custom ops. So if few-or-zero exceptions are encountered in torchbench, we should also add some tests for avoiding C++ exceptions in the common path of torch.compile usage with custom ops.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

albanD commented:
Skipping PR sanity check as the files are auto-generated

bdhirsh commented:
@pytorchbot merge

bdhirsh committed:
Update on "track number of cpp->python exceptions thrown in torch.compile benchmark suite"

cpp->python exceptions can be slow - **especially* if you are running pytorch with infra that tries to symbolize C++ stacktraces.

This PR attempts to:
(1) add a counter every time pybind catches a C++ exception that enters python
(2) track the value of this counter in our benchmark suite, so we can hopefully drive it to zero.

It might be the case that we have very few (or even zero) of these cases showing up in torchbench. Emprically from internal, it also appears that this can show up when torch.compiling custom ops. So if few-or-zero exceptions are encountered in torchbench, we should also add some tests for avoiding C++ exceptions in the common path of torch.compile usage with custom ops.




cc albanD voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge

bdhirsh commented:
@pytorchbot merge

bdhirsh commented:
@pytorchbot merge -i "existing trunk failure"

bdhirsh commented:
@pytorchbot --help

bdhirsh commented:
@pytorchbot merge -i

bdhirsh committed:
Update on "track number of cpp->python exceptions thrown in torch.compile benchmark suite"

cpp->python exceptions can be slow - **especially* if you are running pytorch with infra that tries to symbolize C++ stacktraces.

This PR attempts to:
(1) add a counter every time pybind catches a C++ exception that enters python
(2) track the value of this counter in our benchmark suite, so we can hopefully drive it to zero.

It might be the case that we have very few (or even zero) of these cases showing up in torchbench. Emprically from internal, it also appears that this can show up when torch.compiling custom ops. So if few-or-zero exceptions are encountered in torchbench, we should also add some tests for avoiding C++ exceptions in the common path of torch.compile usage with custom ops.




cc albanD voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bdhirsh commented:
@pytorchbot rebase

bdhirsh committed:
Update

[ghstack-poisoned]

bdhirsh committed:
Update on "track number of cpp->python exceptions thrown in torch.compile benchmark suite"

cpp->python exceptions can be slow - **especially* if you are running pytorch with infra that tries to symbolize C++ stacktraces.

This PR attempts to:
(1) add a counter every time pybind catches a C++ exception that enters python
(2) track the value of this counter in our benchmark suite, so we can hopefully drive it to zero.

It might be the case that we have very few (or even zero) of these cases showing up in torchbench. Emprically from internal, it also appears that this can show up when torch.compiling custom ops. So if few-or-zero exceptions are encountered in torchbench, we should also add some tests for avoiding C++ exceptions in the common path of torch.compile usage with custom ops.




cc albanD voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bdhirsh committed:
Update on "track number of cpp->python exceptions thrown in torch.compile benchmark suite"

cpp->python exceptions can be slow - **especially* if you are running pytorch with infra that tries to symbolize C++ stacktraces.

This PR attempts to:
(1) add a counter every time pybind catches a C++ exception that enters python
(2) track the value of this counter in our benchmark suite, so we can hopefully drive it to zero.

It might be the case that we have very few (or even zero) of these cases showing up in torchbench. Emprically from internal, it also appears that this can show up when torch.compiling custom ops. So if few-or-zero exceptions are encountered in torchbench, we should also add some tests for avoiding C++ exceptions in the common path of torch.compile usage with custom ops.




cc albanD voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bdhirsh committed:
Update on "track number of cpp->python exceptions thrown in torch.compile benchmark suite"

cpp->python exceptions can be slow - **especially* if you are running pytorch with infra that tries to symbolize C++ stacktraces.

This PR attempts to:
(1) add a counter every time pybind catches a C++ exception that enters python
(2) track the value of this counter in our benchmark suite, so we can hopefully drive it to zero.

It might be the case that we have very few (or even zero) of these cases showing up in torchbench. Emprically from internal, it also appears that this can show up when torch.compiling custom ops. So if few-or-zero exceptions are encountered in torchbench, we should also add some tests for avoiding C++ exceptions in the common path of torch.compile usage with custom ops.




cc albanD voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bdhirsh committed:
Update on "track number of cpp->python exceptions thrown in torch.compile benchmark suite"

cpp->python exceptions can be slow - **especially* if you are running pytorch with infra that tries to symbolize C++ stacktraces.

This PR attempts to:
(1) add a counter every time pybind catches a C++ exception that enters python
(2) track the value of this counter in our benchmark suite, so we can hopefully drive it to zero.

It might be the case that we have very few (or even zero) of these cases showing up in torchbench. Emprically from internal, it also appears that this can show up when torch.compiling custom ops. So if few-or-zero exceptions are encountered in torchbench, we should also add some tests for avoiding C++ exceptions in the common path of torch.compile usage with custom ops.




cc albanD voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bdhirsh committed:
Update on "track number of cpp->python exceptions thrown in torch.compile benchmark suite"

cpp->python exceptions can be slow - **especially* if you are running pytorch with infra that tries to symbolize C++ stacktraces.

This PR attempts to:
(1) add a counter every time pybind catches a C++ exception that enters python
(2) track the value of this counter in our benchmark suite, so we can hopefully drive it to zero.

It might be the case that we have very few (or even zero) of these cases showing up in torchbench. Emprically from internal, it also appears that this can show up when torch.compiling custom ops. So if few-or-zero exceptions are encountered in torchbench, we should also add some tests for avoiding C++ exceptions in the common path of torch.compile usage with custom ops.




cc albanD voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bdhirsh committed:
Update on "track number of cpp->python exceptions thrown in torch.compile benchmark suite"

cpp->python exceptions can be slow - **especially* if you are running pytorch with infra that tries to symbolize C++ stacktraces.

This PR attempts to:
(1) add a counter every time pybind catches a C++ exception that enters python
(2) track the value of this counter in our benchmark suite, so we can hopefully drive it to zero.

It might be the case that we have very few (or even zero) of these cases showing up in torchbench. Emprically from internal, it also appears that this can show up when torch.compiling custom ops. So if few-or-zero exceptions are encountered in torchbench, we should also add some tests for avoiding C++ exceptions in the common path of torch.compile usage with custom ops.




cc albanD voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bdhirsh committed:
Update on "track number of cpp->python exceptions thrown in torch.compile benchmark suite"

cpp->python exceptions can be slow - **especially* if you are running pytorch with infra that tries to symbolize C++ stacktraces.

This PR attempts to:
(1) add a counter every time pybind catches a C++ exception that enters python
(2) track the value of this counter in our benchmark suite, so we can hopefully drive it to zero.

It might be the case that we have very few (or even zero) of these cases showing up in torchbench. Emprically from internal, it also appears that this can show up when torch.compiling custom ops. So if few-or-zero exceptions are encountered in torchbench, we should also add some tests for avoiding C++ exceptions in the common path of torch.compile usage with custom ops.




cc albanD voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bdhirsh committed:
Update on "track number of cpp->python exceptions thrown in torch.compile benchmark suite"

cpp->python exceptions can be slow - **especially* if you are running pytorch with infra that tries to symbolize C++ stacktraces.

This PR attempts to:
(1) add a counter every time pybind catches a C++ exception that enters python
(2) track the value of this counter in our benchmark suite, so we can hopefully drive it to zero.

It might be the case that we have very few (or even zero) of these cases showing up in torchbench. Emprically from internal, it also appears that this can show up when torch.compiling custom ops. So if few-or-zero exceptions are encountered in torchbench, we should also add some tests for avoiding C++ exceptions in the common path of torch.compile usage with custom ops.




cc albanD voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bdhirsh committed:
Update on "track number of cpp->python exceptions thrown in torch.compile benchmark suite"

cpp->python exceptions can be slow - **especially* if you are running pytorch with infra that tries to symbolize C++ stacktraces.

This PR attempts to:
(1) add a counter every time pybind catches a C++ exception that enters python
(2) track the value of this counter in our benchmark suite, so we can hopefully drive it to zero.

It might be the case that we have very few (or even zero) of these cases showing up in torchbench. Emprically from internal, it also appears that this can show up when torch.compiling custom ops. So if few-or-zero exceptions are encountered in torchbench, we should also add some tests for avoiding C++ exceptions in the common path of torch.compile usage with custom ops.




cc albanD voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bdhirsh committed:
Update on "track number of cpp->python exceptions thrown in torch.compile benchmark suite"

cpp->python exceptions can be slow - **especially* if you are running pytorch with infra that tries to symbolize C++ stacktraces.

This PR attempts to:
(1) add a counter every time pybind catches a C++ exception that enters python
(2) track the value of this counter in our benchmark suite, so we can hopefully drive it to zero.

It might be the case that we have very few (or even zero) of these cases showing up in torchbench. Emprically from internal, it also appears that this can show up when torch.compiling custom ops. So if few-or-zero exceptions are encountered in torchbench, we should also add some tests for avoiding C++ exceptions in the common path of torch.compile usage with custom ops.




cc albanD voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bdhirsh committed:
Update on "track number of cpp->python exceptions thrown in torch.compile benchmark suite"

cpp->python exceptions can be slow - **especially* if you are running pytorch with infra that tries to symbolize C++ stacktraces.

This PR attempts to:
(1) add a counter every time pybind catches a C++ exception that enters python
(2) track the value of this counter in our benchmark suite, so we can hopefully drive it to zero.

It might be the case that we have very few (or even zero) of these cases showing up in torchbench. Emprically from internal, it also appears that this can show up when torch.compiling custom ops. So if few-or-zero exceptions are encountered in torchbench, we should also add some tests for avoiding C++ exceptions in the common path of torch.compile usage with custom ops.




cc albanD voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bdhirsh committed:
Update on "track number of cpp->python exceptions thrown in torch.compile benchmark suite"

cpp->python exceptions can be slow - **especially* if you are running pytorch with infra that tries to symbolize C++ stacktraces.

This PR attempts to:
(1) add a counter every time pybind catches a C++ exception that enters python
(2) track the value of this counter in our benchmark suite, so we can hopefully drive it to zero.

It might be the case that we have very few (or even zero) of these cases showing up in torchbench. Emprically from internal, it also appears that this can show up when torch.compiling custom ops. So if few-or-zero exceptions are encountered in torchbench, we should also add some tests for avoiding C++ exceptions in the common path of torch.compile usage with custom ops.




cc albanD voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bdhirsh committed:
Update on "track number of cpp->python exceptions thrown in torch.compile benchmark suite"

cpp->python exceptions can be slow - **especially* if you are running pytorch with infra that tries to symbolize C++ stacktraces.

This PR attempts to:
(1) add a counter every time pybind catches a C++ exception that enters python
(2) track the value of this counter in our benchmark suite, so we can hopefully drive it to zero.

It might be the case that we have very few (or even zero) of these cases showing up in torchbench. Emprically from internal, it also appears that this can show up when torch.compiling custom ops. So if few-or-zero exceptions are encountered in torchbench, we should also add some tests for avoiding C++ exceptions in the common path of torch.compile usage with custom ops.




cc albanD voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bdhirsh committed:
Update on "track number of cpp->python exceptions thrown in torch.compile benchmark suite"

cpp->python exceptions can be slow - **especially* if you are running pytorch with infra that tries to symbolize C++ stacktraces.

This PR attempts to:
(1) add a counter every time pybind catches a C++ exception that enters python
(2) track the value of this counter in our benchmark suite, so we can hopefully drive it to zero.

It might be the case that we have very few (or even zero) of these cases showing up in torchbench. Emprically from internal, it also appears that this can show up when torch.compiling custom ops. So if few-or-zero exceptions are encountered in torchbench, we should also add some tests for avoiding C++ exceptions in the common path of torch.compile usage with custom ops.




cc albanD voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bdhirsh committed:
Update on "track number of cpp->python exceptions thrown in torch.compile benchmark suite"

cpp->python exceptions can be slow - **especially* if you are running pytorch with infra that tries to symbolize C++ stacktraces.

This PR attempts to:
(1) add a counter every time pybind catches a C++ exception that enters python
(2) track the value of this counter in our benchmark suite, so we can hopefully drive it to zero.

It might be the case that we have very few (or even zero) of these cases showing up in torchbench. Emprically from internal, it also appears that this can show up when torch.compiling custom ops. So if few-or-zero exceptions are encountered in torchbench, we should also add some tests for avoiding C++ exceptions in the common path of torch.compile usage with custom ops.




cc albanD voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bdhirsh committed:
Update on "track number of cpp->python exceptions thrown in torch.compile benchmark suite"

cpp->python exceptions can be slow - **especially* if you are running pytorch with infra that tries to symbolize C++ stacktraces.

This PR attempts to:
(1) add a counter every time pybind catches a C++ exception that enters python
(2) track the value of this counter in our benchmark suite, so we can hopefully drive it to zero.

It might be the case that we have very few (or even zero) of these cases showing up in torchbench. Emprically from internal, it also appears that this can show up when torch.compiling custom ops. So if few-or-zero exceptions are encountered in torchbench, we should also add some tests for avoiding C++ exceptions in the common path of torch.compile usage with custom ops.




cc albanD voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]
